### ISSUE TITLE
Transect download not working properly

### ISSUE THREAD

**Issue by CrisSevilleja** [BiologicalRecordsCentre/ABLE#489]
At the moment the download of the transect Occurrences file is not working, neither on the Scheme Download page https://butterfly-monitoring.net/downloads nor at the user Download page https://butterfly-monitoring.net/mydata/downloads

The cvs downloaded doesn't have all the information in the first column, there is also data in the second so that makes it complicated to get the file arranged for basic users to use in excel format (change from one column data to the different columns). Also, some users reported that some abundance data is empty where they know there is data for those visits and species. 

Attached is an example from Slovenia:
[occurrences Slovenia 30102022.csv](https://github.com/BiologicalRecordsCentre/ABLE/files/9895610/occurrences.Slovenia.30102022.csv)

Regarding the transect samples file, not all the data is downloaded. I took the Slovenia coordinators as an example with data in transects for 2021 and 2022, and the download of the samples is only for 2021, missing 2022. Spanish coordinator reported similar problems and missing abundance data of recently submitted data.
[samples Slovenia 30102022.csv](https://github.com/BiologicalRecordsCentre/ABLE/files/9895635/samples.Slovenia.30102022.csv)

There are some issues with the download process, and at the moment everyone wants to download the data for the year for making reports. Please resolve this as soon as possible. Thank you

**Comment by CrisSevilleja** [comment]
If possible, add the column "site name" for the transect Occurrences file. At the moment it only appears the site code that coordinators doesn't know by heart, but they recognize better the site name. Thanks

**Comment by DavidRoy** [comment]
@JimBacon can you look at this please

**Comment by JimBacon** [comment]
Hi @CrisSevilleja 

You say

> The cvs downloaded doesn't have all the information in the first column, there is also data in the second so that makes it complicated to get the file arranged for basic users to use in excel format (change from one column data to the different columns). 

You probably know that a CSV file is a Comma-Separated Variable file. Each field of information is separated from the next by a comma. You can see this clearly if you open the downloaded file in a text editor. 

In some countries, where the comma is used as the decimal separator in fractional numbers, Excel tries to be helpful and assumes the semi-colon (;) is the field separator. The consequence is that it make a complete disaster of opening CSV files.

[This article](https://support.microsoft.com/en-us/office/import-or-export-text-txt-or-csv-files-5250ac4c-663c-47ce-937b-339e391393ba) explains how you can correctly open the files. When using the text import wizard, use the following settings.

![image](https://user-images.githubusercontent.com/2545390/199224738-c0d92680-86d7-434e-9d8e-0756a6809194.png)

![image](https://user-images.githubusercontent.com/2545390/199224860-6cbd548d-9332-40e7-870e-78c936fec550.png)

![image](https://user-images.githubusercontent.com/2545390/199225046-f14264bd-1715-4d69-ae74-6f7b7f22d125.png)

In the final step we are specifying the correct format for the date field. All the other fields I have left as 'General'

It was never intended that all the data should be in one column. It is supposed to look like this 

![image](https://user-images.githubusercontent.com/2545390/199225939-1849882a-0a49-4d4e-9f78-0b5e311f5b1e.png)

**Comment by JimBacon** [comment]
The reason that there are records with no abundance count is because, in the configuration of the attributes for the EBMS Transects, there are two with function `sex_stage_count`

![image](https://user-images.githubusercontent.com/2545390/199258756-19d54c6d-e8b7-4d09-9189-f215e69e8e01.png)

The cache_occurrence_nonfunctional.attr_sex_stage_count can only take on one value and, when an Outside Transect Count is supplied, it takes precedence over Abundance count. Outside Transect Count is a termlist value.

The way ElasticSearch had been configured to create the download, only an integeral attr_sex_stage_count  will be returned so, when an Outside Transect Count exists, the abundance count will be missing. Sometimes it is correct, there was no count within the transect, but sometimes the transect count is masked where one exists.

Proposed resolution is to add both attributes to the download so it is clear what the count is.

It might also be beneficial to 
- Remove the sex_stage_count function from Outside Transect Count
- Update the cache to hold the correct value in attr_sex_stage_count where it currently holds an Outside Transect Count.

This might possibly have some unexpected impact on the website so I will not do this for now.

**Comment by JimBacon** [comment]
The  "Download species occurrences from transects" has been updated to include Outside Transect Count and the Transect Name (as requested above).

I was over-complicating the reason for records with no abundance count in my description above (though the problem described does exist). They will simply be occurrences with no abundance count which exist because they have an Outside Transect Count which was previously not visible. 

There still seem to be a very few occurrences (33) with no count either inside or outside the transect.

**Comment by JimBacon** [comment]
Regarding the transect sample data, the Slovenian file attached above contains 45 records with 2 from 2022.
I downloaded it again and now it has 43 records with 5 from 2022. There is some editing in progress it seems.
A query of the database gave identical results and there was no evidence of any inconsistency.
It is to be noted that walks have not been entered in date order. Sorting by date rather than walk ID may help to understand the data.

**Comment by SarahVray** [comment]
Hello, if I may add something, in our case for the export of species occurrences from Luxembourg BMS we also have missing data ([download-63638f0d5c7003.47320082.csv](https://github.com/BiologicalRecordsCentre/ABLE/files/9927344/download-63638f0d5c7003.47320082.csv)). 
- For "Abundance Count", we have 62 NA in 2021 (although the abundance data are visible on the website)
- 72 lines have no information on location (no Transect ID, no Transect Code, no Transect Name, no Section Name) in 2016 and 2017.
- 159 other lines are missing Transect Code (but have a Transect ID and a Transect Name).

I thought these missing values might be relevant to report in this issue.
Thanks a lot in advance for your help,
Best regards,
Sarah Vray

**Comment by CrisSevilleja** [comment]
Thank you Sarah for your information, indeed something is wrong with the download and happen to several coordinators. 

Thanks @JimBacon for explaining the case of the missing abundance, but I don't understand it so well the  sex_stage_count function from Outside Transect Count, sorry. This function, if I understand it well, Outside the transect Count I think it is only Hungary has this option to include data outside the transect, so it shouldn't affect Luxembourg, Spain or Slovenia. Is this what are referring you?

Regarding your answer to the column arrangement of the download data, 

> You probably know that a CSV file is a Comma-Separated Variable file. Each field of information is separated from the next by a comma. You can see this clearly if you open the downloaded file in a text editor.

Yes, I know and that is what I explained to all the users how to arrange their downloaded data from eBMS into a normal excel file by comma-separated variable https://butterfly-monitoring.net/sites/default/files/Pdf/using%20the%20website/How%20to%20open%20a%20CSV.pdf 
However, this is not working well, if I open the [occurrences.Slovenia.30102022.csv](https://github.com/BiologicalRecordsCentre/ABLE/files/9930271/occurrences.Slovenia.30102022.csv) there is an error and not all the data is open in the csv in one column. There should be a missing comma or something, how you can see on the screen-shot the data is in two column, so I cannot replicate the process you explain on how to separate by comma a csv.  I hope I explained myself well.

![image](https://user-images.githubusercontent.com/25052478/199752970-9674a94d-d738-4b6d-bfb2-56329ec1666e.png)

Thank you Jim for trying to solve this. Best wishes,

**Comment by JimBacon** [comment]
Hi @CrisSevilleja 

I'll try to explain the Excel problem another way. When you use Excel to open a CSV file, because of your regional settings, it separates the data in to columns using the semi-colon instead of the comma. There is a semi-colon in our data which is why it ends up in two columns.

There are two solutions mentioned in the article I gave a link to:

1. Start Excel without any file open then follow the instructions to "import a text file by connecting to it"
2. Change the file extension from csv to txt before opening it in the normal way.

Either of these will cause Excel to show the text import wizard so that the file can be opened correctly.

**Comment by chrisvanswaay** [comment]
Reading in R seems to go well:

library(data.table)
dt <- fread("occurrences.Slovenia.30102022.csv")

![image](https://user-images.githubusercontent.com/25886005/200000083-bc4bef3b-5a84-4aa2-840d-1d821f68cc93.png)

**Comment by chrisvanswaay** [comment]
@SarahVray Tried your file too:
library(data.table)
dt <- fread("download-63638f0d5c7003.47320082.csv")

Indeed there are NA's, all at site EBMS:Luxembourg:167 (Givenich 1):

![image](https://user-images.githubusercontent.com/25886005/200001433-2b651e42-6d27-43c4-811f-57c35216a95a.png)

**Comment by xaviermestdagh** [comment]
I highlighted another issue falling under this quite generic issue name...
Here a list of visits without any values for Start time, End time... (column G to K here below)
![image](https://user-images.githubusercontent.com/24294138/200528607-96467cad-f4df-4162-8b95-3dc95b0e0faa.png)

When looking in the "Transect Sample ID" column, all these records are 8 digits ID when others are 7. 
![image](https://user-images.githubusercontent.com/24294138/200530054-39610f3c-d289-4969-a640-e91263f07f49.png)

These values are missing in all views (download export, annual reports, and editing respective visits). 

Do you have any idea on the origin of such missing values?

**Comment by xaviermestdagh** [comment]
Looking to the occurrences export from the Downloads part, there is the "Transect Code" which is not in the sample (visit) export of the same download part. 
On my side, I would be in favour to also have "Transect Code" in the sample (visit) export as well because it was the only site ID available in the annual report (based on which I build all my data management). 
As you can see, this Transect Code is missing for some records, although not missing for records within the same Transect Sample ID...
![image](https://user-images.githubusercontent.com/24294138/200543215-d920d657-5bca-419b-b925-c8e6ec18ede6.png)

**Comment by CrisSevilleja** [comment]
> Hi @CrisSevilleja
> 
> I'll try to explain the Excel problem another way. When you use Excel to open a CSV file, because of your regional settings, it separates the data in to columns using the semi-colon instead of the comma. There is a semi-colon in our data which is why it ends up in two columns.
> 
> There are two solutions mentioned in the article I gave a link to:
> 
> 1. Start Excel without any file open then follow the instructions to "import a text file by connecting to it"
> 2. Change the file extension from csv to txt before opening it in the normal way.
> 
> Either of these will cause Excel to show the text import wizard so that the file can be opened correctly.

Thank you Jim, now I see I can open this file properly. I will see what is the best way to explain to the users to open properly the downloaded files (in general normal users use excel). Regarding the semi-colon, I think it is related to the issue https://github.com/BiologicalRecordsCentre/ABLE/issues/493#issuecomment-1309017292 showing the two sun% that I don't think is useful, maybe better to create another column for the total sun% of the whole transect, avoiding this problem as well.

**Comment by DavidRoy** [comment]
@JimBacon are you still working on this one?

**Comment by JimBacon** [comment]
This issue has become very confusing because it refers to multiple problems.

In summary:

- CSV file doesn't open correctly in Excel. **Resolved**. This is a matter of coercing Excel to recognise the comma as the field separator.
- Slovenia scheme transect samples download missing 2022, data. 2022 data was present. **No fault found**.
- Scheme transect occurrences download missing recent abundance data. 
    - Occurrences with an outside transect count were listed but with no count visible. Added column for outside count. **Resolved**
    - ElasticSearch had incorrect copy of attrs_json. Reindex triggered for Luxembourg records. **Resolved**
    - Identify other records missing abundance in ElasticSearch. **To do**.
- Add Site Name column to scheme transect occurrences download. **Done**.
- Start and end times missing
    - ElasticSearch had incorrect copy of attrs_json. Reindex triggered for affected records. **Resolved**.
    - Other records had no values stored for these attributes. **No action possible**.
- Add Transect Code column to scheme transect samples download. **To do**.
- Transect Code missing from some records in scheme transect occurrences download. **To do**.

In conclusion, work is still needed here.

**Comment by DavidRoy** [comment]
Thanks for the detailed update

@CrisSevilleja can you open new separate issues for the remaining problems.  Then close this one?

**Comment by CrisSevilleja** [comment]
New issues are open for the remaining tasks.
